### ISSUE TITLE
v5: major non-text contrast issues with focus indicators and progress bars

### ISSUE THREAD

**Issue by patrickhlauke** [twbs/bootstrap#31360]
All of the current focus styles for buttons and form controls currently fail WCAG 2.1 SC 1.4.11 Non-text Contrast https://www.w3.org/TR/WCAG21/#non-text-contrast

While focus indication is indeed "visible" in the broad definition of WCAG 2.1 SC 2.4.7 Focus Visible https://www.w3.org/TR/WCAG21/#focus-visible, the focus outline does also need to have sufficient contrast of at least 3:1 against the background under 1.4.11 (if it's the only visual indication of focus). For text inputs/textareas, there's an argument that the blinking caret/cursor (once it has focus) is an additional visual focus indicator, so they're "excused" from having low contrast outline. But for all other cases, the contrast of the outline is currently *way* below 3:1 (in some cases down to 1.4:1).

> ![all button variants with focus indicator](https://user-images.githubusercontent.com/895831/88444255-3873c380-ce14-11ea-9b32-aa0e9ebc9b0f.png)

> ![select example](https://user-images.githubusercontent.com/895831/88444261-43c6ef00-ce14-11ea-9401-638090aa78e7.png)

> ![checkbox example](https://user-images.githubusercontent.com/895831/88444266-4b869380-ce14-11ea-856b-543255c06407.png)

> ![radio button example](https://user-images.githubusercontent.com/895831/88444275-53dece80-ce14-11ea-910c-0fe00c461784.png)

> ![switch toggle example](https://user-images.githubusercontent.com/895831/88444281-5ccfa000-ce14-11ea-8bcf-b49baec535ae.png)

> ![file upload example](https://user-images.githubusercontent.com/895831/88444295-65c07180-ce14-11ea-9bdb-b5a58a0f4287.png)

> ![range example](https://user-images.githubusercontent.com/895831/88444304-6f49d980-ce14-11ea-8479-dd37b0cec9c4.png)

The same is also true for the custom focus indicators used on the main Bootstrap documentation homepage (these seem to be custom-defined for just the documentation).

> ![focused get started button](https://user-images.githubusercontent.com/895831/88444312-77097e00-ce14-11ea-8fcd-53d30d09a99a.png)

> ![focused download button](https://user-images.githubusercontent.com/895831/88444320-812b7c80-ce14-11ea-8618-1b42c030260a.png)

Same issue is also present for pagination

> ![focused pagination example](https://user-images.githubusercontent.com/895831/88444247-2abe3e00-ce14-11ea-8fb1-e62f146ecff1.png)

Additionally, some of the custom background progress bars also fail 1.4.11 - users need to be able to clearly distinguish both the gray background from the white (to see what the total/100% is) and then be able to distinguish the coloured bar from the gray background.

> ![progress bar examples](https://user-images.githubusercontent.com/895831/88444332-8dafd500-ce14-11ea-8446-61b46f13db5b.png)

Note that most automated tools currently don't adequately (or at all) test for 1.4.11.

**Comment by patrickhlauke** [comment]
and for clarity, this issue is also present in bootstrap v4.x, but for that we aimed only at WCAG 2.0. SC 1.4.11 is new to WCAG 2.1

**Comment by mdo** [comment]
What's the best tool to measure the contrast of focus styles? Is this the contrast of the added ring against background/parent element, or against the element that's being focused?

**Comment by mdo** [comment]
And a demo of some outline styles, including Firefox's non-standard `-moz-outline-radius`: https://codepen.io/emdeoh/pen/RwrdOBo?editors=1100

**Comment by ffoodd** [comment]
Kind of duplicate with #29422, isn't it? There're some tools and infos there.

**Comment by patrickhlauke** [comment]
ah, well remembered @ffoodd  yes, this seems to cover the same ground as #29422 when it comes to focus styles, and adds the problem of progress bars as well (which is arguably separate, but still falls under 1.4.11)

@mdo i mean without even needing tools...you just check the contrast ratio of the outline/shadow colour against the background page. for the handful of focus styles we have, it's reasonably quick to do. i just do it manually with https://github.com/ThePacielloGroup/CCAe/ 

also, while the `-moz-outline-radius` thing is interesting, note that i'm not comparing the contrast of the outline against the button/thing itself, but the contrast of the outline against the page background. it's not about the outline appearing distinct somehow from the button, but the outline actually being visible in and of itself against the page.

**Comment by ffoodd** [comment]
@mdo for a fast overview, [Aditus' Button Contrast Checker](https://www.aditus.io/button-contrast-checker/) is pretty good (and as Patrick said, it measures contrast ratio between focus indicator and background page.

[You and @MartijnCuppens came up with different solutions for this (comment)](https://github.com/twbs/bootstrap/issues/29422#issuecomment-540172066), basically using the same color for button shadows and backgrounds, with a gap between shadow and border to ensure visibility.

This is what I did in Boosted with `outline` — since our buttons aren't rounded, `outline` and `outline-offset` are great.

**Comment by ffoodd** [comment]
Closing this in favor of #29422
